libraries/AP_HAL/RCOutput.cpp

```diff
@@ -0,0 +1,46 @@
+#include "AP_HAL.h"
+
+extern const AP_HAL::HAL &hal;
+
+// helper function for implementation of get_output_mode_banner
+const char* AP_HAL::RCOutput::get_output_mode_string(enum output_mode out_mode) const
+{
+    // convert mode to string
+    switch (out_mode) {
+    case MODE_PWM_NONE:
+    default:
+        return "None";
+    case MODE_PWM_NORMAL:
+        return "PWM";
+    case MODE_PWM_ONESHOT:
+        return "OneS";
+    case MODE_PWM_ONESHOT125:
+        return "OS125";
+    case MODE_PWM_BRUSHED:
+        return "Brush";
+    case MODE_PWM_DSHOT150:
+        return "DS150";
+    case MODE_PWM_DSHOT300:
+        return "DS300";
+    case MODE_PWM_DSHOT600:
+        return "DS600";
+    case MODE_PWM_DSHOT1200:
+        return "DS1200";
+    case MODE_NEOPIXEL:
+        return "NeoP";
+    }
+}
+
+// convert output mode to string.  helper function for implementation of get_output_mode_banner
+void AP_HAL::RCOutput::append_to_banner(char banner_msg[], uint8_t banner_msg_len, output_mode out_mode, uint8_t low_ch, uint8_t high_ch) const
```

Comment:
This function doesn't actually know it's appending anything to anything.  It just knows that it;s been given a buffer to shove stuff into.

I can't offer anything better than `banner_string_bit_for_mode_and_channels`

Reply:
well, not exactly actually.  If you look closely at the snprintf you'll see that both it's source and destination are banner_msg.  So it really is appending.

This appending using sprintf is actually the bit of this PR that I'm unsure about,  It seems to work but it does worry me a little to have the source and destination be the same.

Reply:
Ah.  Why not have append_to_banner return the output from `snprintf`?  The caller can keep track of where the next stuff is to go into the buffer and pass `&banner_msg[banner_msg_len]` into this function.  It will need to worry about clamping that `snprintf` output value in that case (or just whinge and break out of the loop if it runs out of space in the buffer).  `snprintf` returns the number of bytes that would have been required if it had successfully output everything.

Reply:
@peterbarker, thanks again for the feedback.. before I go and do that work though, is there any advantage?  Is there anything dangerous or incorrect with how it is in this PR at the moment?

Reply:
@rmackay9 I don't think it dangerous or incorrect with our *current* implementation of snprintf.  However, I have looked at optimising https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_HAL/utility/print_vprintf.cpp#L380 to do a memcpy rather than bytewise assignment, and this code would break then.  I've not found any standard for snprintf for covering the buffer-and-arguments-overlap case.

Just not something we've done.  We do a fair bit of passing-in-pointers-to-bits-of-a-buffer, 'though.

The current mechanism is obviously also hideously inefficient, too :-)  For a once-off call it's shrug-worthy, but it is n^2 rather than n.

Reply:
@peterbarker, I haven't made it more efficient (probably less) but I've made it safer (I think) by taking a backup of the original string before doing the snprintf.

Reply:
The other option is to just `asprintf` a new string each time and delete the passed in previous one. This uses memory on the heap so you have to delete it, but saves you the copy aside, and with the longer status text support we just gained this would actually cope with that situation reasonably for fully documenting the assignment. I'm okay with the allocation/deletion stuff, since this would only be done once during startup.

Reply:
I'm hopeful that what I've done is safe and simple enough that everyone finds it acceptable.  I don't think it's worth optimising too much in this function that is so rarely called.

So, if anyone objects to how it's done, please tell me clearly that you want it changed.

Reply:
I won't hold this up for that.  I don't think any of us see actual bugs here, just alternative implementations - and I don't see anyone volunteering to own new bugs :-)

Reply:
@WickedShell, @peterbarker, cool.  really thanks very much for the reviews.  reviews are super valuable.. so really appreciate it.